featurebyte/common/utils.py

```diff
@@ -420,3 +424,132 @@ def _repr_html_(self) -> str:
+                    },
+                )
+                key_column.appendChild(doc.createTextNode(key))
+                row.appendChild(key_column)
+
+                # populate value column
+                if isinstance(value, dict):
+                    # process dictionaries recursively
+                    value_elem = doc.createElement("div")
+                    _set_element_style(value_elem, {"border": "0", "padding": "0", "margin": "0"})
+                    _populate_html_elem(
+                        data=value, doc=doc, elem=value_elem, html_content=html_content
+                    )
+                elif isinstance(value, list) and isinstance(value[0], dict):
+                    # list of dictionaries
+                    value_elem = doc.createElement("div")
+                    _set_element_style(value_elem, {"border": "0", "padding": "0", "margin": "0"})
+                    df_key = str(ObjectId())
+                    html_content[df_key] = pd.DataFrame(value).to_html()
+                    value_elem.appendChild(doc.createTextNode(f"{{{df_key}}}"))
+                else:
+                    # nicer datetime formatting
+                    try:
+                        value = datetime.fromisoformat(value).strftime("%Y-%m-%d %H:%M:%S")
+                    except (ValueError, TypeError):
+                        pass
+                    value_elem = doc.createTextNode(str(value))
+
+                val_column = doc.createElement("td")
+                _set_element_style(val_column, {"width": "80%", "text-align": "left"})
+                val_column.appendChild(value_elem)
+                row.appendChild(val_column)
+
+            elem.appendChild(table)
+
+        # create html document
+        impl = getDOMImplementation()
+        assert impl
+        doc_type = impl.createDocumentType(
+            "html",
+            "-//W3C//DTD XHTML 1.0 Strict//EN",
+            "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd",
+        )
+        doc = impl.createDocument("http://www.w3.org/1999/xhtml", "html", doc_type)
+        doc_elem = doc.documentElement
+
+        # add title for the table
+        data = copy.deepcopy(self)
+        class_name = re.sub(r"([A-Z]+[a-z]+)", r" \1", self.class_name).strip()
+        title_div = doc.createElement("div")
+        title_div.appendChild(doc.createTextNode(class_name))
+        _set_element_style(
+            elem=title_div,
+            style={
+                "font-weight": "bold",
+                "text-align": "center",
+                "border-bottom": "1px solid black",
+                "width": "100%",
+                "padding-top": "20px",
+            },
+        )
+        doc_elem.appendChild(title_div)
+
+        # add info table
+        html_content: Dict[str, str] = {}
+        _populate_html_elem(data=data, doc=doc, elem=doc_elem, html_content=html_content)
+        html = str(doc.toprettyxml())
+        return html.format(**html_content)
```

Comment:
This part could be useful for rendering other type of object in the jupyter notebook environment. We could move this to external scope (outside this function scope) later when we want to reuse it. 👍

Reply:
That makes sense, will move it.